pkg/logger/logger_cluster.go

```diff
@@ -0,0 +1,156 @@
+type Cluster interface {
+	//GetLoggerPipeline returns the cluster instance the Cluster send the coupled ChannelMessages to
+	GetLoggerPipeline() Pipeline
+
+	//GetChannelProvider returns the channel provider this Cluster uses
+	GetChannelProvider() ChannelProvider
+
+	//StartListening enables the Cluster instance to listen to the channel found in Cluster#getChannelProvider
+	StartListening()
+
+	//Flushes the cluster content to the pipeline
+	Flush()
+
+	//GetWaitGroup returns the waitgroup instance the cluster uses to identify if it is done running
+	GetWaitGroup() *sync.WaitGroup
+}
+
+//simpleCluster is a default implementation of the Cluster interface
+type simpleCluster struct {
+	pipe            Pipeline
+	channel         ChannelProvider
+	clusterDuration time.Duration
+	waitGroup       *sync.WaitGroup
+	messageCache    []ChannelMessage
+	nextCacheFlush  time.Time
+}
+
+//GetLoggerPipeline returns the upstream pipeline the Cluster sends all its data to
+func (s *simpleCluster) GetLoggerPipeline() Pipeline {
+	return s.pipe
+}
+
+//GetChannelProvider returns the channel provider the Cluster wi
+func (s *simpleCluster) GetChannelProvider() ChannelProvider {
+	return s.channel
+}
+
+//StartListening enables the Cluster instance to listen to the channel found in Cluster#getChannelProvider
+//This will block all code executed, so use it in a go routine
+func (s *simpleCluster) StartListening() {
+	s.GetWaitGroup().Add(1)
+
+	for incomingMessage := range s.GetChannelProvider().GetChannel() {
+		if s.timedFlushNeeded() {
+			if s.isLoggerMessageCached(incomingMessage.GetLogger().GetID()) { //The message is not flushed this release
+				s.Flush()
+				s.resetFlushTimer()
+
+				s.appendMessage(incomingMessage)
+			} else { //The message is flushed in this release
+				s.appendMessage(incomingMessage)
+
+				s.Flush()
+				s.resetFlushTimer()
+			}
+		} else { //Time wise we don't need to flush
+			if s.isLoggerMessageCached(incomingMessage.GetLogger().GetID()) { //Overwriting logger message, gotta flush
+				s.Flush()
+				s.resetFlushTimer()
+			}
+
+			s.appendMessage(incomingMessage) //Add message to either this or new release, depending on prev if
+		}
+	}
+
+	s.Flush()
+	s.GetWaitGroup().Done()
+}
+
+//Flush flushes all cached messages to the pipeline
+func (s *simpleCluster) Flush() {
+	if len(s.messageCache) < 1 {
+		return
+	}
+
+	s.GetLoggerPipeline().Write(s.messageCache)
+	s.resetMessageCache()
+}
+
+//GetWaitGroup returns the waitgroup used to signal if the listening is running at the momement
+func (s *simpleCluster) GetWaitGroup() *sync.WaitGroup {
+	return s.waitGroup
+}
+
+//resetMessageCache simply clears the message cache
+func (s *simpleCluster) resetMessageCache() {
+	s.messageCache = nil
+	//signal the garbage collector to remove eventually cached array structure under it: https://stackoverflow.com/questions/16971741/how-do-you-clear-a-slice-in-go
```

Comment:
If this is the explanation why you set it to `nil`, I would suggest to move it before the statement to avoid confusion.